Lib/importlib/_bootstrap_external.py

```diff
@@ -1711,6 +1711,57 @@ def __repr__(self):
         return f'FileFinder({self.path!r})'
 
 
+class AppleFrameworkLoader(ExtensionFileLoader):
+    """A loader for modules that have been packaged as frameworks for
+    compatibility with Apple's iOS App Store policies.
+    """
+    def create_module(self, spec):
+        mod = super().create_module(spec)
+        if spec.loader_state.origfile is not None:
+            mod.__file__ = spec.loader_state.origfile
+        return mod
+
+
+class AppleFrameworkFinder:
+    """A finder for modules that have been packaged as Apple Frameworks
+    for compatibility with Apple's App Store policies.
+
+    See AppleFrameworkLoader for details.
+    """
+    def __init__(self, frameworks_path):
+        self.frameworks_path = frameworks_path
+
+    def find_spec(self, fullname, paths, target=None):
+        name = fullname.rpartition(".")[-1]
+
+        for extension in EXTENSION_SUFFIXES:
+            dylib_file = _path_join(
+                self.frameworks_path,
+                f"{fullname}.framework",
+                f"{name}{extension}",
+            )
+            _bootstrap._verbose_message("Looking for Apple Framework dylib {}", dylib_file)
+
+            dylib_exists = _path_isfile(dylib_file)
+            if dylib_exists:
+                origfile = None
+                if paths:
+                    for parent_path in paths:
```

Comment:
If I understand correctly, when importing a top-level module, `paths` is None, so the `__file__` attribute will not be rewritten. 

In fact, I don't think it *could* be rewritten with the information available. If an app has both the standard library and some other source of binary modules on `sys.path` (whether it's `app_packages`, `site_packages` or anything else), then the current framework naming convention makes it impossible to know which path entry the "original" location was.

I don't have any immediate solution for this. Maybe the framework naming convention needs to be rethought.

Reply:
Such as by adding some additional dotted segments to the start, making the framework name effectively a path relative to dirname(sys.executable)?

Reply:
This would also ensure there are no clashes with any other frameworks that happen to be in the app.

Reply:
The new loader implementation I've just pushed should resolve this. It leaves a `.fwork` breadcrumb which is used to anchor the discovery of the module, which allows us to canonically know where a binary came from, without any guesswork, even if it's in `lib-dynload` or somewhere else at a root of a `sys.path` entry.